### PR TITLE
Amend `unifyTypeList` to account for subtyping

### PR DESCRIPTION
**Pull request by ansley**
Stack from [ghstack](https://github.com/ezyang/ghstack):
* **#59226 Amend `unifyTypeList` to account for subtyping**

Because the number of types in our types system is small, we've been able to get away with an incorrect (albeit fast) version of `unifyTypeList` that's basically the same as calling `std::accumulate` with `unifyTypes`.  This doesn't account for the case that two types cannot themselves be unified, but there exists some mutual parent type later in the container `elements`. (E.g. `T2` <: `T1`, `T3` <: `T1`, `elements` = {`T2`, `T3`, `T1`}. The real supertype should be `T1`, not `c10::nullopt`.)

Although this PR changes `unifyTypeList` from `O(N)` to `O(N^2)`, we won't see any practical performance degradation. If our type system stays roughly the same size as it is today, `N` will be so small that a nested loop won't matter. If we introduce polymorphic types in the next year or so and our type system expands greatly, we would need the `O(N^2)` algorithm anyway. So why not go ahead and change it?

Differential Revision: [D28807203](https://our.internmc.facebook.com/intern/diff/D28807203)

### CODE REVIEW

**Comment by ansley** [comment]
@ansley has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D28807203).

File: aten/src/ATen/core/type.cpp [hunk 1]
```diff
@@ -346,21 +347,54 @@ c10::optional<TypePtr> unifyTypeList(
...
+      // completely new type (e.g. `unifyTypes(T, None)` produces
+      // `Optional[T]`)
+      if (maybe_parent && type->isSubtypeOf(*maybe_parent)) {
+        seen.erase(type);
+        seen.insert(*maybe_parent);
+      };
+      if (maybe_parent && comparison->isSubtypeOf(*maybe_parent)) {
+        seen.erase(comparison);
+        seen.insert(*maybe_parent);
+      };
```
Comment (SplitInfinity) [inline aten/src/ATen/core/type.cpp:387]:
```suggestion
      if (maybe_parent) {
        seen.insert(*maybe_parent);
        
        if (type->isSubtypeOf(*maybe_parent)) {
          seen.erase(type);
        };
      
        if (comparison->isSubtypeOf(*maybe_parent)) {
          seen.erase(comparison);
        };
      }
```

File: test/jit/test_typing.py [hunk 2]
```diff
@@ -200,12 +200,14 @@ def fn2(x):
         self.checkScript(fn, [])
         self.checkScript(fn2, (torch.ones(2, 2),))
 
-        with self.assertRaisesRegex(RuntimeError, "Could not unify"):
+        with self.assertRaisesRegex(RuntimeError, "Lists must contain "
+                                    "only a single type"):
```
Comment (SplitInfinity) [inline test/jit/test_typing.py:204]: Why did these error messages changes? Because of extra calls to `unifyTypes1 that weren't being made before?

**Review by SplitInfinity — COMMENTED** [review]
Is there a concrete example or use case in which `unifyTypeList` has returned the wrong result? It would be nice to add a unit test containing that example so that we can confirm that this new implementation works as intended.

File: aten/src/ATen/core/type.cpp [hunk 3]
```diff
@@ -346,21 +347,54 @@ c10::optional<TypePtr> unifyTypeList(
...
+      // possible supertypes
+      if (*type == *comparison) {
+        continue;
+      }
+
+      auto maybe_parent = unifyTypes(type, comparison, /*default_to_any=*/false);
+
+      // This logic covers the case in which `unifyTypes` returns a
+      // completely new type (e.g. `unifyTypes(T, None)` produces
+      // `Optional[T]`)
```
Comment (SplitInfinity) [inline aten/src/ATen/core/type.cpp:379]:
```suggestion
      // `Optional[T]`).
```

**Review by tugsbayasgalan — COMMENTED** [review]
This is pretty cool. But could you add more unit tests where the previous change would fail while the new one would pass?

File: test/jit/test_typing.py (hunk 2 above)
Comment (tugsbayasgalan) [reply test/jit/test_typing.py:204]: +1

**Comment by ansley** [comment]
@SplitInfinity @tugsbayasgalan

I should have clarified this earlier, but there isn't currently a use case for this (hence, no test cases). However, there probably will be a use case for it in the future, and Yanan has suggested I fix it as part of some other PRs (e.g. #57137). Basil had a similar suggestion when he reviewed my project on dynamic types. So, I'm opening a separate PR since there are several pending PRs I'd like to use this algorithm in (again, not because it's necessary, rather as a response to comments)

File: aten/src/ATen/core/jit_type.h [hunk 4]
```diff
@@ -1572,6 +1572,8 @@ TORCH_API c10::optional<TypePtr> unifyTypes(
     const TypePtr& t2,
     bool default_to_any = false);
 
+// Get the smallest common supertype (excluding `Any`) of all the values
+// in the list. If there is no common supertype, return `c10::nullopt`
```
[the remark this answers is not in the document]
Comment (ansley) [reply aten/src/ATen/core/jit_type.h:1576]: no :)

**Comment by SplitInfinity** [comment]
> @SplitInfinity @tugsbayasgalan
> 
> I should have clarified this earlier, but there isn't currently a use case for this (hence, no test cases). However, there probably will be a use case for it in the future, and Yanan has suggested I fix it as part of some other PRs (e.g. #57137). Basil had a similar suggestion when he reviewed my project on dynamic types. So, I'm opening a separate PR since there are several pending PRs I'd like to use this algorithm in (again, not because it's necessary, rather as a response to comments)

That's fine, but I still think there should be a unit test for this new behaviour. If such a test cannot be added until later, I would prefer that the PR(s) that enable(s) a test to be added be stacked on top and landed together.

File: aten/src/ATen/core/type.cpp [hunk 5]
```diff
@@ -346,21 +347,54 @@ c10::optional<TypePtr> unifyTypeList(
...
+  // `std::accumulate` to account for the case that two types
+  // cannot themselves be unified, but there exists some mutual
+  // parent type later in `elements`. (E.g. T2 <: T1, T3 <: T1,
+  // `elements` = {T2, T3, T1}. The real supertype should be `T1`,
+  // not `c10::nullopt`)
+
+  // Set of possible parent types
+  std::unordered_set<TypePtr, torch::jit::HashType, torch::jit::EqualType> seen;
+
+  for (const TypePtr type : elements) {
```
[the remark this answers is not in the document]
Comment (ansley) [reply aten/src/ATen/core/type.cpp:361]: Is there a reason we want to use `auto`? I think it only makes sense when the type is long or unwieldy, e.g. a lambda

[the remark this answers is not in the document]
Comment (SplitInfinity) [reply aten/src/ATen/core/type.cpp:361]: If the type of `elements` ever changes, you won't have to update this for loop if it uses `auto`.

File: aten/src/ATen/core/type.cpp [hunk 6]
```diff
@@ -346,21 +347,54 @@ c10::optional<TypePtr> unifyTypeList(
...
+    seen.insert(type);
+
+    // We'll modify `seen` during this loop, so we need to make a copy
+    // of the values that we want to iterate over
+    std::vector<TypePtr> to_iterate{seen.begin(), seen.end()};
+
+    for (const TypePtr comparison : to_iterate) {
+      // Don't bother compare if it's already in our set of
```
[the remark this answers is not in the document]
Comment (ansley) [reply aten/src/ATen/core/type.cpp:369]: impressed

**Review by tugsbayasgalan — APPROVED** [review]

**Comment by ansley** [comment]
@ansley has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D28807203).

**Review by eellison — COMMENTED** [review]
Could you add a test case that failed before but works now ?

**Comment by ansley** [comment]
Yanan and I discussed this, and we'd like to land it now. I have two PRs open in which my reviewers have brought up the issue of forward compatibility with subtyping relationships. So, even though the new `unifyTypeList` algorithm isn't relevant to our type system now, it's relevant to some PR comments now. I plan to go back and add additional tests when it's possible to do so.

**Review by eellison — CHANGES_REQUESTED** [review]
would like to verify correctness and have a test case that shows that it was incorrect beforrer

**Comment by eellison** [comment]
I don't think this is right - unifyTypes is the place for the fix, not unifyTypeList. We should not be making changes to core type & jit infrastructure without test cases.

**Comment by eellison** [comment]
> This doesn't account for the case that two types cannot themselves be unified, but there exists some mutual parent 

This should be fixed in `unifyTypes` - it should always return the most exact supertype, which in this case is the mutual parent

**Comment by ansley** [comment]
@eellison Please see example in PR description. This problem only occurs when unifying a list of types using a naïve `accumulate`-style algorithm

**Comment by eellison** [comment]
@ansley look at my most recent comment

**Comment by ansley** [comment]
@eellison Good point. I think you're right here. If it's okay with @gmagogsfm, I'll close this and ship the pending PRs

**Pull request closed without merging**